content/posts/2021/test-python-functions/index.md

```diff
@@ -0,0 +1,387 @@
+
+### Add the first tests
+Let's start with some simple tests:
+
+```sh
+$ touch calc/handler_test.py
+```
+
+then add the following test cases for a couple of our happy paths
+
+```py
+import calc.handler as h
+
+
+class TestParsing:
+    def test_operation_addition(self):
+        req = '{"op": "+", "var1": "1.0", "var2": 0}'
+        resp, code = h.handle(req)
+        assert code == 200
+        assert resp["value"] == 1.0
+
+    def test_operation_multiplication(self):
+        req = '{"op": "*", "var1": "100.01", "var2": 1}'
+        resp, code = h.handle(req)
+        assert code == 200
+        assert resp["value"] == 100.01
+```
+
+At this point, we have followed normal conventions for pytest, by default, pytest will look for files that match the pattern `*_test.py` and then the run the functions that  match `test_*` (including matching methods if the class is named `Test*`, which is what we did above).
+
+Now, we can run our test and see all of the errors in their wonderful glory
+
+```sh
+$ pytest
+========================================= test session starts =========================================
+platform linux -- Python 3.8.8, pytest-6.2.2, py-1.10.0, pluggy-0.13.1
+rootdir: /home/lucas/code/openfaas/sandbox/pytest-sample
+collected 2 items
+
+calc/handler_test.py FF                                                                         [100%]
+
+============================================== FAILURES ===============================================
+_________________________________ TestParsing.test_operation_addition _________________________________
+
+self = <calc.handler_test.TestParsing object at 0x7fc8029ad820>
+
+    def test_operation_addition(self):
+        req = '{"op": "+", "var1": "1.0", "var2": 0}'
+>       resp, code = h.handle(req)
+E       ValueError: too many values to unpack (expected 2)
+
+calc/handler_test.py:49: ValueError
+______________________________ TestParsing.test_operation_multiplication ______________________________
+
+self = <calc.handler_test.TestParsing object at 0x7fc8029add30>
+
+    def test_operation_multiplication(self):
+        req = '{"op": "^", "var1": "2", "var2": -2}'
+>       resp, code = h.handle(req)
+E       ValueError: too many values to unpack (expected 2)
+
+calc/handler_test.py:73: ValueError
+======================================= short test summary info =======================================
+FAILED calc/handler_test.py::TestParsing::test_operation_addition - ValueError: too many values to u...
+FAILED calc/handler_test.py::TestParsing::test_operation_multiplication - ValueError: too many value...
+========================================== 2 failed in 0.05s ==========================================
+```
+
+In this case we get a value error because our handler function only returns a string, not a dictionary and a status code like the test expects.  These test cases  are expecting that the handler returns something that looks like this
```

Comment:
```suggestion
In this case we get a value error because our handler function only returns a string, not a dictionary and a status code like the test expects.  These test cases expect the handler to return something that looks like this
```